### DISCUSSION TITLE
How to pass custom header from router to subgraph

### DISCUSSION THREAD

**Discussion by Gladys-LM** [apollographql/router#466] [General]
I am trying to do a POC using apollo router + API-One, both deployed to k8s. API-One is a monolith graph and is expecting a custom header. I tried to use Postman to set the header for the post request to router endpoint, the traffic went from router to API-One fine but got error response because the custom header wasn't passed from router to API-One. Tried to tweak router configuration but didn't work. Any tips on how I can pass the custom header from router to API-One? Thanks.

**Comment by o0Ignition0o** [comment]
Hi, thanks for opening this discussion!

We're about to land our plugins system, which will allow you to cover this usecase. I'll get back to you on that once we merge it on main

**Reply by dakshas** [reply]
Hi, is there any update on this ?

**Reply by o0Ignition0o** [reply]
Indeed!

we know have fully fledged [header propagation rules which you can configure with yml](https://www.apollographql.com/docs/router/configuration/header-propagation)

please let me know if it covers your use case!

**Comment by andrew-kolesnikov** [comment]
Those who want to **generate a custom header for subgraph inside the router**, instead of **forwarding from router downstream into subgraph**, will need a custom plugin:

`    fn subgraph_service(
        &mut self,
        _name: &str,
        service: BoxService<SubgraphRequest, SubgraphResponse, BoxError>,
    ) -> BoxService<SubgraphRequest, SubgraphResponse, BoxError> {
        ServiceBuilder::new()
            .map_request(|mut req: SubgraphRequest| {
                req.subgraph_request.headers_mut().insert("foo", HeaderValue::from_str("bar").unwrap());
                req
            })
            .service(service)
            .boxed()
    }`

**Reply by bnjjj** [reply]
Maybe I don't understand your need but it's already covered by the configuration file by using `insert` https://www.apollographql.com/docs/router/configuration/header-propagation/#insert

**Reply by andrew-kolesnikov** [reply]
> Maybe I don't understand your need but it's already covered by the configuration file by using `insert` https://www.apollographql.com/docs/router/configuration/header-propagation/#insert

There are many common scenarios for which `insert` is not suitable, e.g. pass a `JWT` with custom claims generated _inside the router_ downstream into a subgraph. The custom plugin suggested above looks like the only available option.

**Reply by bnjjj** [reply]
OK my bad, regarding your example I thought it was a static content

**Reply by garypen** [reply]
If writing a rust plugin seems like a challenge, perhaps [write a rhai script ](https://github.com/apollographql/router/tree/main/examples/cookies-to-headers) to create and propagate.
